neuralmonkey/decoders/transformer.py

```diff
@@ -143,101 +157,122 @@ def embedded_train_inputs(self) -> tf.Tensor:
+        """Create the feed-forward network sublayer."""
+
+        # Layer normalization
+        normalized_input = layer_norm(layer_input)
+
+        # Feed-forward network hidden layer + ReLU
+        ff_hidden = tf.layers.dense(
+            normalized_input, self.ff_hidden_size, activation=tf.nn.relu,
+            name="hidden_state")
+
+        # Apply dropout on the activations
+        ff_hidden = dropout(ff_hidden, self.dropout_keep_prob, self.train_mode)
+
+        # Feed-forward output projection
+        ff_output = tf.layers.dense(ff_hidden, self.dimension, name="output")
+
+        # Apply dropout on the output projection
+        ff_output = dropout(ff_output, self.dropout_keep_prob, self.train_mode)
+
+        # Add residual connections
+        return ff_output + layer_input
 
     def layer(self, level: int, inputs: tf.Tensor,
               mask: tf.Tensor) -> TransformerLayer:
-        # Recursive implementation. Outputs of the zeroth layer are the inputs
+        # Recursive implementation. Outputs of the zeroth layer
+        # are the inputs
+
         if level == 0:
-            norm_inputs = tf.contrib.layers.layer_norm(
-                inputs, begin_norm_axis=2)
-            return TransformerLayer(norm_inputs, mask)
+            return TransformerLayer(inputs, mask)
 
         # Compute the outputs of the previous layer
         prev_layer = self.layer(level - 1, inputs, mask)
 
-        # Run self-attention
-        self_context = self.masked_self_attention(level, prev_layer)
-
-        # Residual connections + layer normalization
-        encoder_queries = tf.contrib.layers.layer_norm(
-            self_context + prev_layer.temporal_states, begin_norm_axis=2)
+        with tf.variable_scope("layer_{}".format(level - 1)):
 
-        # Attend to the encoder
-        encoder_context = self.encoder_attention(level, encoder_queries)
+            with tf.variable_scope("self_attention", reuse=tf.AUTO_REUSE):
+                self_context = self.self_attention_sublayer(prev_layer)
 
-        # Residual connections + layer normalization
-        ff_input = tf.contrib.layers.layer_norm(
-            encoder_context + encoder_queries, begin_norm_axis=2)
+            with tf.variable_scope("encdec_attention", reuse=tf.AUTO_REUSE):
+                encoder_context = self.encoder_attention_sublayer(self_context)
 
-        # Feed-forward network hidden layer + ReLU + dropout
-        ff_hidden = tf.layers.dense(
-            ff_input, self.ff_hidden_size, activation=tf.nn.relu,
-            name="ff_hidden_{}".format(level))
-        ff_hidden_drop = dropout(
-            ff_hidden, self.dropout_keep_prob, self.train_mode)
-
-        # Feed-forward output projection + dropout
-        ff_output = tf.layers.dense(
-            ff_hidden_drop, self.dimension, name="ff_out_{}".format(level))
-        ff_output = dropout(ff_output, self.dropout_keep_prob, self.train_mode)
+            with tf.variable_scope("feedforward"):
```

Comment:
co tady, proč tu neni to `tf.AUTO_REUSE`? Myslíte že by se sem hodilo?

Reply:
Nebo naopak že se nehodí do těch nahoře?

Reply:
AUTO_REUSE jsem vyhodil